### PR TITLE
refactor neg controller: processService, add unittest

### PR DESCRIPTION
**Pull request by cadmuxe**

### CODE REVIEW

File: pkg/neg/controller.go [hunk 1]
```diff
@@ -325,69 +325,31 @@ func (c *Controller) processService(key string) error {
 	if service == nil {
 		return fmt.Errorf("cannot convert to Service (%T)", obj)
 	}
-	negAnnotation, foundNEGAnnotation, err := annotations.FromService(service).NEGAnnotation()
-	if err != nil {
-		return err
-	}
 
 	portInfoMap := make(negtypes.PortInfoMap)
```
Comment (freehan) [inline pkg/neg/controller.go]: s/portInfoMap/svcPortInfoMap

Comment (freehan) [reply pkg/neg/controller.go]: also add a comment:

svcPortInfoMap contains NEG related port info for the service

File: pkg/neg/controller.go [hunk 2]
```diff
@@ -325,69 +325,31 @@ func (c *Controller) processService(key string) error {
...
-				return fmt.Errorf("failed to merge service ports referenced by Istio:DestinationRule (%v): %v", servicePortInfoMap, err)
-			}
-		}
+	destinationRulesPortInfoMap, err := c.mergeCSMPortInfoMap(namespace, name, service, &portInfoMap)
+	if err != nil {
+		return err
 	}
 
-	if needNeg {
+	if len(portInfoMap) != 0 {
```
Comment (freehan) [inline pkg/neg/controller.go]: if len(svcPortInfoMap) != 0 || len(destinationRulesPortInfoMap) != 0

File: pkg/neg/controller.go [hunk 3]
```diff
@@ -325,69 +325,31 @@ func (c *Controller) processService(key string) error {
...
-	// Only enable for L7-ILB for now to limit possible issues
-	// TODO(shance): investigate enabling this for all ingresses
-	if flags.F.EnableL7Ilb && c.defaultBackendService.ID.Service.String() == key {
-		portInfoMap = c.defaultBackendServicePortInfoMap()
-		needNeg = len(portInfoMap) > 0
-	} else {
-		if foundNEGAnnotation && negAnnotation != nil && negAnnotation.NEGEnabled() {
-			needNeg = true
-		}
+	if err := c.mergeDefaultBackendServicePortInfoMap(key, &portInfoMap); err != nil {
```
Comment (freehan) [inline pkg/neg/controller.go]: no need to pass the pointer of portInfoMap.

map in golang is already pass as reference

File: pkg/neg/controller.go [hunk 4]
```diff
@@ -325,69 +325,31 @@ func (c *Controller) processService(key string) error {
...
+		return err
+	}
+	if err := c.mergeIngressPortInfo(service, types.NamespacedName{Namespace: namespace, Name: name}, &portInfoMap); err != nil {
```
Comment (freehan) [inline pkg/neg/controller.go]: no need for pointer

File: pkg/neg/controller.go [hunk 5]
```diff
@@ -325,69 +325,31 @@ func (c *Controller) processService(key string) error {
...
+		return err
+	}
+	if err := c.mergeStandaloneNEGsPortInfo(service, types.NamespacedName{Namespace: namespace, Name: name}, &portInfoMap); err != nil {
```
Comment (freehan) [inline pkg/neg/controller.go]: same here

File: pkg/neg/controller.go [hunk 6]
```diff
@@ -400,7 +362,15 @@ func (c *Controller) processService(key string) error {
 	return c.syncNegStatusAnnotation(namespace, name, make(negtypes.PortInfoMap))
 }
 
-func (c *Controller) mergeIngressPortInfo(negAnnotation *annotations.NegAnnotation, service *apiv1.Service, name types.NamespacedName, portInfoMap *negtypes.PortInfoMap) error {
+func (c *Controller) mergeIngressPortInfo(service *apiv1.Service, name types.NamespacedName, portInfoMap *negtypes.PortInfoMap) error {
```
Comment (freehan) [inline pkg/neg/controller.go]: add comments

File: pkg/neg/controller.go [hunk 7]
```diff
@@ -411,7 +381,17 @@ func (c *Controller) mergeIngressPortInfo(negAnnotation *annotations.NegAnnotati
 			return fmt.Errorf("failed to merge service ports referenced by ingress (%v): %v", ingressPortInfoMap, err)
 		}
 	}
+	return nil
+}
 
+func (c *Controller) mergeStandaloneNEGsPortInfo(service *apiv1.Service, name types.NamespacedName, portInfoMap *negtypes.PortInfoMap) error {
```
Comment (freehan) [inline pkg/neg/controller.go]: add comments

File: pkg/neg/controller.go [hunk 8]
```diff
@@ -325,69 +325,31 @@ func (c *Controller) processService(key string) error {
...
-		} else if contains(c.csmServiceNEGSkipNamespaces, namespace) {
-			klog.Infof("Skip NEG creation for services in namespace: %s", namespace)
-		} else {
-			needNeg = true
-			servicePortInfoMap := negtypes.NewPortInfoMap(namespace, name, servicePorts, c.namer, true)
-			if err := portInfoMap.Merge(servicePortInfoMap); err != nil {
-				return fmt.Errorf("failed to merge service ports referenced by Istio:DestinationRule (%v): %v", servicePortInfoMap, err)
-			}
-		}
+	destinationRulesPortInfoMap, err := c.mergeCSMPortInfoMap(namespace, name, service, &portInfoMap)
```
Comment (freehan) [inline pkg/neg/controller.go]: mergeCSMPortInfoMap -> getCSMPortInfoMap  

you can return both svcPortInfo and destinationRulesPortInfoMap from this function

File: pkg/neg/controller.go [hunk 9]
```diff
@@ -325,69 +325,31 @@ func (c *Controller) processService(key string) error {
...
-		}
 		if err = c.syncNegStatusAnnotation(namespace, name, portInfoMap); err != nil {
 			return err
 		}
 		// Merge destinationRule related NEG after the Service NEGStatus Sync, we don't want DR related NEG status go into service.
-		if err := portInfoMap.Merge(csmPortInfoMap); err != nil {
-			return fmt.Errorf("failed to merge service ports referenced by Istio:DestinationRule (%v): %v", csmPortInfoMap, err)
+		if err := portInfoMap.Merge(destinationRulesPortInfoMap); err != nil {
+			return fmt.Errorf("failed to merge service ports referenced by Istio:DestinationRule (%v): %v", destinationRulesPortInfoMap, err)
 		}
```
Comment (freehan) [inline pkg/neg/controller.go]: had a local variable to hold the combined svcPortInfoMap and destinationRulesPortInfoMap

File: pkg/neg/controller_test.go [hunk 10]
```diff
@@ -581,15 +590,89 @@ func TestDefaultBackendServicePortInfoMap(t *testing.T) {
...
-
-			result := controller.defaultBackendServicePortInfoMap()
+			result := make(negtypes.PortInfoMap)
+			controller.mergeDefaultBackendServicePortInfoMap(controller.defaultBackendService.ID.Service.String(), &result)
 			if !reflect.DeepEqual(tc.want, result) {
 				t.Fatalf("got %+v, want %+v", result, tc.want)
 			}
 		})
 	}
 }
```
Comment (freehan) [inline pkg/neg/controller_test.go]: add a unit test for CSM NEG case similar to `TestNewNEGService`.

It should have combination of dummy services and destinationrules. And validate: 
1. The syncers are correct
2. The annotation are correct


This is sort of a e2e test for processService in the CSM mode.

File: pkg/neg/controller.go [hunk 11]
```diff
@@ -432,20 +412,65 @@ func (c *Controller) mergeIngressPortInfo(negAnnotation *annotations.NegAnnotati
...
+		} else {
+			servicePortInfoMap := negtypes.NewPortInfoMap(namespace, name, servicePorts, c.namer, false)
+			if err := portInfoMap.Merge(servicePortInfoMap); err != nil {
+				return nil, fmt.Errorf("failed to merge service ports referenced by Istio:DestinationRule (%v): %v", servicePortInfoMap, err)
+			}
+		}
+		return destinationRulesPortInfoMap, nil
 	}
-	return make(negtypes.PortInfoMap)
+	return nil, nil
```
Comment (freehan) [inline pkg/neg/controller.go]: it should return empty map instead of nil in case of nil pointer

File: pkg/neg/controller.go (hunk 1 above)
Comment (cadmuxe) [reply pkg/neg/controller.go]: Done

File: pkg/neg/controller.go (hunk 2 above)
Comment (cadmuxe) [reply pkg/neg/controller.go]: done

File: pkg/neg/controller.go (hunk 3 above)
Comment (cadmuxe) [reply pkg/neg/controller.go]: Thanks.

File: pkg/neg/controller.go (hunk 4 above)
Comment (cadmuxe) [reply pkg/neg/controller.go]: updated.

File: pkg/neg/controller.go (hunk 9 above)
Comment (cadmuxe) [reply pkg/neg/controller.go]: What's the purpose to have the combined PortInfoMap?

**Comment by cadmuxe** [comment]
Still working on the unittest.

**Comment by freehan** [comment]
LGTM

Please squash commits. Just have one commit for vendoring, one for the refactoring+unit test.


/approve

**Pull request merged**